File: src/main/scala/llmTests/a201902a/scala/gpt4/WorkflowImpl.scala
package llmTests.a201902a.scala.gpt4

import java.util.{LinkedHashSet, LinkedList}
import llmTests.a201902a.scala.{Workflow, WorkflowsFactory}

import scala.collection.mutable
import scala.jdk.CollectionConverters._

private class WorkflowImpl[T](initialTasks: Set[T], dependencyMap: Map[T, Set[T]]) extends Workflow[T] {
  private val tasks: mutable.Set[T] = initialTasks.to(mutable.Set)
  private val tasksToDo = mutable.Set.from(initialTasks)
  private val completedTasks = mutable.Set.empty[T]

  override def getTasks: java.util.Set[T] = tasks.asJava
  override def getNextTasksToDo: java.util.Set[T] = tasksToDo.asJava

  override def doTask(t: T): Unit = {
    if (!tasks.contains(t)) throw new IllegalArgumentException("Unknown task")
    if (!tasksToDo.contains(t)) throw new IllegalStateException("Cannot execute this task yet")
    tasksToDo.remove(t)
    completedTasks.add(t)

    dependencyMap.getOrElse(t, Set()).foreach { dependentTask =>
      if (!completedTasks.exists(dependencyMap(dependentTask).contains)) {
        tasksToDo.add(dependentTask)
      }
    }
  }

  override def isCompleted: Boolean = tasks.isEmpty || tasks == completedTasks
}

File: src/main/scala/llmTests/a201902a/java/gpt35turbo/WorkflowImpl.java
package llmTests.a201902a.java.gpt35turbo;

import llmTests.a201902a.java.Workflow;

import java.util.*;

class WorkflowImpl<T> implements Workflow<T> {

    private Set<T> tasks;
    private Set<T> completedTasks;
    private Map<T, Set<T>> dependencies;

    public WorkflowImpl(Set<T> tasks, Set<T> completedTasks, Map<T, Set<T>> dependencies) {
        this.tasks = tasks;
        this.completedTasks = completedTasks;
        this.dependencies = dependencies;
    }

    public Set<T> getTasks() {
        return tasks;
    }

    public boolean isCompleted() {
        return tasks.isEmpty();
    }

    public Set<T> getNextTasksToDo() {
        Set<T> nextTasks = new HashSet<>();
        for (T task : tasks) {
            boolean canBeExecuted = true;
            Set<T> taskDependencies = dependencies.get(task);
            if (taskDependencies != null) {
                for (T dependency : taskDependencies) {
                    if (!completedTasks.contains(dependency)) {
                        canBeExecuted = false;
                    }
                }
            }
            if (canBeExecuted) {
                nextTasks.add(task);
            }
        }
        return nextTasks;
    }

    public void doTask(T task) {
        if (!tasks.contains(task)) {
            throw new IllegalArgumentException();
        }
        for (T dependency : dependencies.keySet()) {
            dependencies.get(dependency).remove(task);
        }
        tasks.remove(task);
        completedTasks.add(task);
    }
}